src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -1031,8 +1083,29 @@ def validate_global_constraints(self):
                 "could not find an ancestor OMPParallelDirective node")
 
         self._validate_single_loop()
+        self._validate_collapse_value()
 
-        super(OMPDoDirective, self).validate_global_constraints()
+        super().validate_global_constraints()
+
+    def _validate_collapse_value(self):
+        '''
+        Checks that if there is a collapse clause, there must be as many
+        immediately nested loops as the collapse value.
+
+        :raises GenerationError: if this OMPLoopDirective has a collapse \
+            clause but it doesn't have the expected number of nested Loops.
+        '''
+        if self._collapse:
+            cursor = self.dir_body.children[0]
+            for depth in range(self._collapse):
+                if not isinstance(cursor, Loop):
+                    raise GenerationError(
+                        f"{type(self).__name__} must have as many immediately "
+                        f"nested loops as the collapse clause specifies but "
+                        f"'{self}' has a collapse={self._collapse} and the "
+                        f"nested statement at depth {depth} is a "
+                        f"{type(cursor).__name__} rather than a Loop.")
+                cursor = cursor.loop_body.children[0]
```

Comment:
What about if the loops aren't perfectly nested? e.g.:

    do ji = 1, jpi
      do jj = 1, jpj
      ...
      end do
      var(ji) = x
    end do

Reply:
Ups, now I also check that the loop is alone in the previous loop body.
And I improved the test cases as suggested.